### PR TITLE
feat(eval): dataset-level evaluators + classification schemas, samples, e2e

### PR DESCRIPTION
**Pull request by ajay-kesavan**
## Summary

Completes the classification evaluator feature shipped in #1397 by adding the three pieces that PR didn't carry:

1. **Generated type schemas** — `BinaryClassificationEvaluator.json` and `MulticlassClassificationEvaluator.json` under `packages/uipath/src/uipath/eval/evaluators_types/`, produced by `python -m uipath.eval.evaluators_types.generate_types`. These are the machine-readable schemas external tooling (Flow UI evaluator picker, `uip maestro flow eval`) uses to know each evaluator's config / criteria / justification shape.

2. **Sample projects** under `packages/uipath/samples/`:
   - `binary_classification_agent/` — rule-based spam/ham classifier wired to the binary classification evaluator with `metric_type=precision`. Eval set is designed so 4/5 datapoints pass but precision is 2/3 because of one deliberate false positive — demonstrates the dataset-level metric diverging from a simple per-row pass rate.
   - `multiclass_classification_simple/` — rule-based 3-class router (payments / support / spam) wired to the multiclass classification evaluator with `averaging=macro`. Eval set forces a misroute that hurts both payments precision and support recall, giving macro F1 = (0.8 + 0.8 + 1.0) / 3.

3. **End-to-end test** at `packages/uipath/tests/cli/eval/test_classification_samples_e2e.py` — loads each sample's eval set, wires its `main.py` into a stand-in runtime, calls `evaluate()`, and asserts both the per-row scores and the aggregated metric produced by `reduce_scores`. Locks in the dataset-level math.

## Why split this PR

PR #1397 added the Python implementation and registered the new evaluator type IDs (`uipath-binary-classification`, `uipath-multiclass-classification`) in the coded-evaluator discriminator, but didn't regenerate the JSON type files or add a runnable example. Without these the evaluators are merged-in-name-only.

## Test plan

- [x] `pytest tests/cli/eval/test_classification_samples_e2e.py` — both samples pass
- [x] `ruff check tests/cli/eval/test_classification_samples_e2e.py` — clean
- [x] `ruff format --check` — clean
- [x] `cat packages/uipath/src/uipath/eval/evaluators_types/BinaryClassificationEvaluator.json` exposes `positive_class`, `metric_type`, `f_value` in `evaluatorConfigSchema.properties`
- [x] `cat packages/uipath/src/uipath/eval/evaluators_types/MulticlassClassificationEvaluator.json` exposes `classes`, `averaging`, `metric_type`, `f_value`
- [ ] CI passes

## Related PRs

- #1664 — companion PR that refreshes the 11 *unrelated* stale schemas in the same directory (split out for review hygiene; no functional overlap with this PR).
- UiPath/cli#2128 — TypeScript-side flow-tool registry entries that wire these evaluators into the Flow UI evaluator picker.

---
Co-Authored-By: Claude Opus 4.7 (1M context) <noreply@anthropic.com>

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by ajay-kesavan** [comment]
Superseded by #1674 (ClassifierEvaluator). The schema/sample work here was replaced by the simpler single-evaluator approach.

**Comment by ajay-kesavan** [comment]
@chatgpt-codex-connector — please run an **adversarial code review** on this PR using the methodology below. Treat the code as a rival's, not a colleague's. Start from "this change is broken" and hunt evidence. If you can't find a real bug after genuinely trying, a positive review is fine — but the search has to happen first.

### Three non-negotiable promises
1. **Cross-check every invoked interface.** When A calls B, open B and read its signature. Prop drift is the #1 source of silent bugs in AI-generated code.
2. **Read test bodies, not test titles.** PR descriptions over-claim, commit messages over-claim, and tests get renamed to match the new implementation rather than the original behavior.
3. **Severity is factual, not emotional.** "Loses user data" = 🔴 even when the author is senior. Typo = 🔵 even when it annoys you.

### 7-step method (skip none)
1. **Inventory the change** — PR metadata, files, commits, full diff. Look for: additions ≈ deletions + "no behavior change" claim (behavior probably changed); commit titled "address review" (open the file it claims to forward into; verify the prop actually lands); mergeable=CONFLICTING (surface it).
2. **Read the full diff top-to-bottom in one pass.** Bugs live at seams between unrelated files in the same diff.
3. **Cross-check every consumer of every new prop / new component / new DI seam.** Grep call sites; confirm every prop both *exists* on the receiver's type AND is *used* (not silently dropped in one branch of an `if`).
4. **Trace data flow end-to-end.** user input → onChange → state → validation → persist → reload. Flag silent drops, empty-string coercion to `{}`/`null`, stale-closure `useState(propDefault)` that never updates, missing memoization on every-render stringification.
5. **Diff the deleted code against what replaced it.** Every `-` line is an unverified claim. Verify: vanished behavioral features (truncation, maxLines, special cases, fallbacks), edge-case handling (empty string, null, arrays), a11y attrs (id, htmlFor, aria-\*), layout-critical CSS classes.
6. **Sanity-check the tests.** Read bodies. Ask "does this verify NEW BEHAVIOR or just the NEW IMPLEMENTATION?" Renames like "renders X" → "renders Y" often launder a regression. Any DI seam MUST have a test that exercises the injected path; fallback-only tests prove nothing.
7. **Classify with severity tiers and post inline anchored comments.**

### Severity tiers
- 🔴 **Critical — blocking** — data loss, crash, security, broken contract, a11y regression
- 🟠 **High — should fix before merge** — structural regression, missing coverage for new code path, error-handling gap
- 🟡 **Medium — quality nit** — smell, redundancy, naming, small perf with evidence
- 🔵 **Low — observation** — style, docs, process

Decision rule: if the PR cannot merge without introducing the bug you describe → 🔴. Merges but degrades UX → 🟠. Below that → 🟡 / 🔵.

### Output format
Per finding: `file:line` citation + concrete fix. State issues as facts ("This loses user input" — not "may potentially"). Criticism first, positives last. Drop softening qualifiers (no "might", "perhaps", "consider"). Adversarial ≠ abusive: attack the code, name the pattern, never the person.

### Red flags — stop and re-read
- "This looks fine" → you haven't cross-checked invoked interfaces
- "Tests pass so it's good" → tests assert the new impl, not the old behavior
- "Description says no behavior change" → diff the deleted lines
- "Small PR, quick review" → small PRs hide big bugs in prop drops
- "AI-generated, probably clean" → AI code is exactly where prop drift hides
### This PR's specific surface area
- `aggregators[]` embedded in `BinaryClassificationEvaluatorConfig` and `MulticlassClassificationEvaluatorConfig`
- New `model_validator`s cross-check `positive_class` / `classes` containment AND `averaging` / `f_value` divergence between evaluator-level and aggregator-level fields when types match
- `reduce_scores` retained on Binary/Multiclass for the per-evaluator headline — the dataset evaluator framework runs additively; verify the two paths agree on identical inputs
- `BaseEvaluatorJustification.try_from` classmethod collapsed 3 duplicate coercion blocks
- Sample JSONs (`binary_classification_agent/`, `multiclass_classification_simple/`) carry new `aggregators[]` arrays
- E2E test asserts specific aggregated scores (binary: P=R=5/6, F1=0.8). Read the multiclass assertions too — they assert specific fscore numbers

File: packages/uipath/src/uipath/eval/evaluators_types/BinaryClassificationEvaluator.json [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+          "f-score"
+        ],
+        "title": "Metric Type",
+        "type": "string"
+      },
+      "f_value": {
+        "default": 1.0,
+        "title": "F Value",
+        "type": "number"
+      }
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/evaluators_types/BinaryClassificationEvaluator.json]: Regenerated `BinaryClassificationEvaluator.json` and `MulticlassClassificationEvaluator.json` from the updated pydantic models. Both schemas now declare `aggregators` plus the `PrecisionAggregatorSpec` / `RecallAggregatorSpec` / `FScoreAggregatorSpec` discriminated union under `$defs`. The 11 other schemas under `evaluators_types/` have pre-existing main drift unrelated to this PR (e.g., `target_output_key` widened to `str | list[str]`, line-by-line fields added on Contains/ExactMatch) — leaving those for a separate sync PR to keep this one's scope tight. `363855d4`.

File: packages/uipath/src/uipath/eval/evaluators/base_dataset_evaluator.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+
+    def __init__(self, spec: SpecT, source_evaluator: str) -> None:
+        """Store the aggregator spec and the source evaluator name."""
+        self.spec = spec
+        self.source_evaluator = source_evaluator
+
+    @property
+    def name(self) -> str:
+        """Stable key for this dataset evaluator's result in the output map."""
+        return f"{self.source_evaluator}.{self.spec.type}"
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/evaluators/base_dataset_evaluator.py]: Fixed in #1669's `4d6afcca` and merged through to this branch. `compute_dataset_evaluator_results` (runtime.py:268) now goes through `_dataset_result_key(source, spec, disambiguate)` which uses the simple `{source}.{type}` key in the common single-aggregator case and appends `.{averaging}` (plus `.fb{f_value}` for fscore) when the same type repeats on the same source. Two new tests cover the duplicate-precision and duplicate-fscore paths. Note: #1663's own H2 model_validator additionally rejects mismatched `averaging` / `f_value` between evaluator-level and aggregator-level when types match, providing a second layer of defense at parse time.

**Comment by ajay-kesavan** [comment]
@chatgpt-codex-connector — please run an **adversarial code review** on this PR.

Treat the code as a rival's, not a colleague's. Start from "this change is broken" and hunt evidence. Look specifically for:

- **Logic bugs / off-by-one / null-deref** in new code paths
- **Concurrency / determinism** issues in Temporal workflows or Python worker
- **Contract drift** between C#/Python payloads (nullability, enum values, discriminators)
- **Silent failures / swallowed exceptions** that mask real errors
- **Security**: injection surfaces, tenant leaks, auth-bypass, unbounded loops from untrusted input
- **Test gaps**: assertions that would pass a broken implementation, mocks that swallow the interesting case
- **Missing invariants**: state transitions that can leave partial data, empty-collection edge cases

If you can't find a real bug after genuinely trying, a positive review is fine — but the review must show you *tried* to break it: list what you attacked, what didn't yield.

Anchor every finding to `file:line`. Rank P0/P1/P2/P3.

**Comment by ajay-kesavan** [comment]
## Review Response — automated review round 1

Ran deep-review + ponytail. Consolidation done, CI passing, no blockers.

### Deferred to follow-up
| Reviewer | Sev | Finding | Plan |
|---|---|---|---|
| deep-review F1 | P2 | Silent overwrite when two aggregators share (type, averaging, f_value) but differ in `classes` | Real bug. Fix is a ~10-line `model_validator` rejecting duplicate `(type, averaging, f_value, sorted(classes))` signatures. Landing in follow-up. |
| deep-review F2 | P2 | `MulticlassClassificationEvaluator.reduce_scores` silently drops OOV, diverges from dataset-eval path | Simplest fix per review: delete the override, inherit `GenericBaseEvaluator.reduce_scores`'s simple average. Aggregator path is the source of truth for the metric-aware number. Landing in follow-up. |
| deep-review F3 | P2 | Asymmetric averaging validation binary vs multiclass | Symmetry fix: reject `spec.averaging != "micro"` when the aggregator's `type` matches the binary evaluator's `metric_type`. One validator addition. Landing in follow-up. |
| deep-review F4 | P2 | Confusion-matrix orientation `[predicted][expected]` not self-describing on the wire | Wire-format concern — resolved by publishing schema doc + adding a top-level `axes` field on the dataset-eval result. Not blocking this PR. |
| deep-review F5-F7 | P3 | Bare-except in `try_from`, legacy `typing.Union/Iterable`, hard-coded isinstance tuple | Acknowledged. F5 and F7 fold naturally into the ponytail cleanups below. F6 is a lint sweep against the whole repo, out of scope here. |
| ponytail (11 items) | mostly YAGNI | See `/reviews/ponytail-review-1663.md` for the batch. Top: delete single-caller `dataset_evaluator_factory.py`, inline `Generic[SpecT]` ABC, drop `_micro_metric` re-derivation (=accuracy for multiclass), extract shared `_prf(...)` helper. | Landing as one focused cleanup PR after this merges — net −125 to −155 LOC. Keeping the shape-changes out of this feature PR. |

_Full review outputs in session scratchpad._

File: packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    for r in results:
+        j = BaseEvaluatorJustification.try_from(r.details)
+        if j is None:
+            n_skipped += 1
+            continue
+        exp = j.expected.lower()
+        act = j.actual.lower()
+        if exp not in index_of or act not in index_of:
+            n_skipped += 1
+            continue
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py]: Real. Deep-review flagged the same as F2 (P2) — codex escalated to P1 because OOV silently changes the P/R/F number without any annotation. Fix per the Review Response plan: delete the `reduce_scores` override on `MulticlassClassificationEvaluator` and inherit `GenericBaseEvaluator`'s simple average, letting the aggregator path be the metric-aware source of truth for classification numbers. Landing in the follow-up cleanup PR against this same feature branch.

File: packages/uipath/src/uipath/eval/runtime/runtime.py [hunk 4]
```diff
@@ -202,6 +210,97 @@ def compute_evaluator_scores(
...
+    """
+    results_by_evaluator: defaultdict[str, list[EvaluationResultDto]] = defaultdict(
+        list
+    )
+    for eval_run_result in evaluation_set_results:
+        for eval_run_result_dto in eval_run_result.evaluation_run_results:
+            if eval_run_result_dto.is_line_result:
+                continue
+            results_by_evaluator[eval_run_result_dto.evaluator_name].append(
+                eval_run_result_dto.result
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/runtime/runtime.py]: Real. Bundling with deep-review F1 (silent overwrite when two aggregators share type/averaging/f_value differ in classes) — same failure surface. Fix: extend `_dataset_result_key` to include a hash of the sorted `classes` tuple AND reject duplicate `(type, averaging, f_value, classes)` at the `model_validator(mode='after')` layer so runtime dedup can drop the disambiguation branch. Landing in follow-up cleanup PR.

File: packages/uipath/src/uipath/eval/runtime/runtime.py [hunk 5]
```diff
@@ -202,6 +210,97 @@ def compute_evaluator_scores(
...
+
+    Uses ``{source}.{type}`` for unique-type aggregators, and appends
+    ``.{averaging}`` (plus ``.fb{f_value}`` for fscore) when the same type
+    appears more than once on the same source.
+    """
+    if not disambiguate:
+        return f"{source_name}.{spec.type}"
+    if isinstance(spec, FScoreAggregatorSpec):
+        return f"{source_name}.{spec.type}.{spec.averaging}.fb{spec.f_value}"
+    return f"{source_name}.{spec.type}.{spec.averaging}"
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/runtime/runtime.py]: Same underlying issue as the previous thread (dedupe) — the disambiguated key doesn't carry classes. Fixed together by the model_validator + `_dataset_result_key` change described above. Follow-up PR.

File: packages/uipath/src/uipath/eval/evaluators/base_dataset_evaluator.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+
+class BaseDatasetEvaluator(ABC, Generic[SpecT]):
+    """Abstract base for dataset-level evaluators.
+
+    Constructed from an :class:`AggregatorSpec` and the name of the source
+    per-datapoint evaluator whose results this aggregator consumes. The
+    dataset evaluator's "name" used for result keying is derived from
+    ``"{source_evaluator}.{spec.type}"`` so two aggregators on the same source
+    don't collide.
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/evaluators/base_dataset_evaluator.py]: Overlaps with codex's earlier thread on aggregator-key uniqueness — same root cause. Fix landing in the follow-up cleanup PR: dedupe validator at model_validator(mode='after') on classes+averaging+f_value, letting _dataset_result_key shed its disambiguation branch.

File: packages/uipath/src/uipath/eval/evaluators/multiclass_classification_evaluator.py [hunk 7]
```diff
@@ -43,6 +55,67 @@ class MulticlassClassificationEvaluatorConfig(
...
+                raise ValueError(
+                    f"Aggregator '{spec.type}' on evaluator '{self.name}' "
+                    f"declares classes={spec.classes!r} but the evaluator's "
+                    f"classes={self.classes!r} include {sorted(missing)!r} "
+                    "that the aggregator does not. Aggregators must cover "
+                    "the evaluator's full class space."
+                )
+            if spec.type == evaluator_aggregator_type:
+                if spec.averaging != self.averaging:
+                    raise ValueError(
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/evaluators/multiclass_classification_evaluator.py]: Real. The current _validate_aggregators_against_evaluator_config requires spec.classes as superset of evaluator.classes by design — an aggregator can compute macro over a wider vocabulary than the evaluator's per-datapoint classifier. Zero-support classes yield defined-but-zero P/R for that class in macro averaging. Follow-up PR: emit a ValidationWarning when spec.classes strictly supersets evaluator.classes so users are aware the extra classes will pull macro averages down.

File: packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+    tp: int
+    tn: int
+    fp: int
+    fn: int
+    support: int
+    value: float
+
+
+class ClassificationDetails(BaseModel):
+    """Structured details payload emitted by every classification evaluator."""
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py]: Correct — ClassificationDetails is a dataset-aggregator-only shape. Per-datapoint BinaryClassificationEvaluator.evaluate() and MulticlassClassificationEvaluator.evaluate() emit a 1.0/0.0 score with structured Justification containing predicted/actual labels; only the aggregator's reduce_scores() synthesizes the confusion matrix + per-class metrics. Docstring will clarify this in the follow-up PR.

File: packages/uipath/src/uipath/eval/runtime/runtime.py [hunk 9]
```diff
@@ -202,6 +210,97 @@ def compute_evaluator_scores(
...
+        # simple key shape in the common case.
+        type_counts: dict[str, int] = defaultdict(int)
+        for spec in config.aggregators:
+            type_counts[spec.type] += 1
+        for spec in config.aggregators:
+            dataset_evaluator = build_dataset_evaluator(spec, source_name)
+            key = _dataset_result_key(source_name, spec, type_counts[spec.type] > 1)
+            dataset_results[key] = EvaluationResultDto.from_evaluation_result(
+                dataset_evaluator.evaluate(source_results)
+            )
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/runtime/runtime.py]: Same finding as codex's aggregator-key threads above — the disambiguated key doesn't carry classes. Fix in follow-up cleanup PR via the model_validator + _dataset_result_key extension already scoped.

File: packages/uipath/src/uipath/eval/runtime/runtime.py [hunk 10]
```diff
@@ -202,6 +210,97 @@ def compute_evaluator_scores(
...
+        evaluators: Per-datapoint evaluator instances that ran during this eval
+            set. Their configs may carry ``aggregators`` lists.
+
+    Returns:
+        Dict mapping ``"{evaluator_name}.{aggregator_type}"`` to the run-level
+        EvaluationResultDto. When the same aggregator ``type`` appears more
+        than once on a source (e.g. macro+micro precision), each variant is
+        disambiguated as ``"{evaluator_name}.{type}.{averaging}"`` and, for
+        fscore, with the ``f_value`` suffix (``"...fbN"``), so a duplicate
+        type never overwrites a previous result. Aggregators whose source
```
[the remark this answers is not in the document]
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/runtime/runtime.py]: Same finding — key disambiguation branch drops classes. Fixed in the follow-up cleanup PR.

**Comment by ajay-kesavan** [comment]
## Review round — post-main-merge (deep + adversarial + ponytail)

Merged `origin/main` (→ v2.13.14, head `1ea11cde`); merge verified clean (536 evaluator tests pass, schema regenerates byte-identical, no aggregator paths dropped).

**Fixed**
- **P2 (deep/adversarial):** `compute_dataset_evaluator_results` now dedups per-datapoint results by `(datapoint, evaluator)` before aggregating, mirroring `compute_evaluator_scores`. The partial-failure path emitted a details-less zero DTO for an evaluator that already produced a real result, and an external retry/resume could re-feed a datapoint — both inflated `nTotal`/`nSkipped` and risked double-counting matrix pairs. Parseable-justification DTO wins. +regression test.
- **Ponytail:** removed the dead `precision==recall==0` early return in `_f_beta` (the `denom==0` branch returns the identical value; β>0, both terms non-negative).

**Reviewed, no change (defensible):**
- 'Count OOV predictions as misses' — matches sklearn's `labels=`-restricted semantics, the k×k contract has no 'other' bucket by design, and the drop is documented + surfaced via `nSkipped`/`nTotal`. Known, transparent limitation of the fixed-vocabulary contract.

Stale round-1 Codex items (BinaryClassificationEvaluator.json schema, `name` property) reference files already removed in earlier rounds.

File: packages/uipath/src/uipath/eval/evaluators/exact_match_evaluator.py [hunk 11]
```diff
@@ -20,6 +23,59 @@ class ExactMatchEvaluatorConfig(OutputEvaluatorConfig[OutputEvaluationCriteria])
...
+                "``classes`` list. Set ``classes`` to the label vocabulary the "
+                "aggregators should compute Precision/Recall/F-score over."
+            )
+        if self.line_by_line_evaluator:
+            raise ValueError(
+                f"ExactMatch evaluator '{self.name}': aggregators are not "
+                "supported with line_by_line_evaluator — per-line results carry "
+                "no expected/actual labels, so every datapoint would be skipped."
+            )
+        if self.case_sensitive:
```
Comment (Chibionos) [inline packages/uipath/src/uipath/eval/evaluators/exact_match_evaluator.py:60]: **P1 — correctness:** `_validate_aggregators` rejects `case_sensitive` and `line_by_line_evaluator` alongside `aggregators` (good — both would silently corrupt the matrix), but doesn't reject `negated`.

`ExactMatchEvaluator.evaluate()` (same file, line ~128) flips `is_exact_match` when `negated=True`, but the `validated_justification` it emits always carries the *raw* `expected`/`actual` strings — the negation only affects `score`, never the justification. `classification_dataset_evaluators.py`'s `_build_confusion` rebuilds the matrix straight from that raw expected/actual pair via `BaseEvaluatorJustification.try_from(r.details)`, with no awareness of `negated` at all. So a negated evaluator's per-datapoint score (correctly inverted) and its dataset-level precision/recall/F-score (built from the non-inverted labels) tell contradictory stories — a "correct" negated match (score=1.0, i.e. actual intentionally != expected) still lands on the confusion-matrix diagonal as if it were a true positive.

```python
        if self.negated:
            raise ValueError(
                f"ExactMatch evaluator '{self.name}': aggregators are not "
                "supported with negated — dataset metrics are computed from "
                "expected/actual labels and would report the non-negated "
                "classification result."
            )
```

(Surfaced by the mandatory Codex recheck on this review; verified independently against the PR's head commit `0d76fe8` — `negated` is defined at exact_match_evaluator.py:25, used at line 128, and absent from `_validate_aggregators`.)

File: packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+    n_skipped = 0
+
+    for r in results:
+        j = BaseEvaluatorJustification.try_from(r.details)
+        if j is None:
+            n_skipped += 1
+            continue
+        exp = j.expected.lower()
+        act = j.actual.lower()
+        if exp not in index_of or act not in index_of:
```
Comment (Chibionos) [inline packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py]: **P1 — correctness:** when `exp`/`act` aren't in the configured `classes` vocabulary, the datapoint is dropped from the matrix entirely (`n_skipped += 1`) rather than counted as a miss for the true class.

This is the identical concern chatgpt-codex-connector raised as P1 on this PR's now-removed `MulticlassClassificationEvaluator` ("Count out-of-vocab predictions as misses" — a 9×`support→support` + 1×`support→fish` run reported `support` recall/F1 as 1.0 with `n_skipped=1` instead of penalizing the false negative). @ajay-kesavan replied "Real" there and said the fix (dropping the `MulticlassClassificationEvaluator.reduce_scores` override) was "landing in the follow-up cleanup PR." That class is gone now, but the aggregator math it was about got absorbed into this file's `_build_confusion`, and the skip-not-miss behavior is unchanged here — still locked in by `test_out_of_vocab_labels_are_skipped` in the new test file. `n_scored`/`n_skipped` are exposed in `ClassificationDetails`, so a consumer who checks them can catch it, but the headline `score` (precision/recall/F-score) alone still reads inflated for anyone who doesn't.

Worth a quick confirmation: is the "count OOV as a miss" fix still intended to land against *this* code path (in the promised follow-up), or was the decision to keep OOV as skip-and-report intentional for the redesigned aggregator model? Either is defensible, but it wasn't decided against the current file, only the old one.

(Surfaced by the mandatory Codex recheck; cross-checked against the existing review thread on the deleted `MulticlassClassificationEvaluator` file and the current test suite.)

**Review by Chibionos — COMMENTED** [review]
## /deep-review — PR #1663 · feat(eval): dataset-level evaluators + classification schemas, samples, e2e

**Branch:** `feat/classification-evaluator-types` → `main` | **Size:** +1628/−7 across 13 files

### Verdict
Comment — two P1 correctness gaps below need a decision before merge. Separately (see standalone PR comment): the title/description/test-plan describe a design this PR no longer ships.

### Slop meter
🟡 low (18/100 slop, 0/100 tunnel-vision) — hits are false positives (idiomatic `Union`/`Annotated` typing, current-style `ConfigDict`), reviewed and dismissed.

### Codex recheck
**Verdict:** amend — Codex agreed with the reviewed findings and surfaced one additional P1 (F1 below), now incorporated and independently verified against the PR's head commit.

### Findings index (bodies inline below)

| # | Sev | File:Line | Headline |
|---|---|---|---|
| F1 | P1 | packages/uipath/src/uipath/eval/evaluators/exact_match_evaluator.py:60 | `negated=True` is allowed with `aggregators`, but the confusion matrix is built from raw expected/actual labels and ignores the negation |
| F2 | P1 | packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py:137 | Out-of-vocab predictions are silently skipped rather than counted as misses — same issue codex flagged as P1 on the now-removed MulticlassClassificationEvaluator; the author called it "Real" there but the current aggregator code still has it |

**Comment by Chibionos** [comment]
**P1 — process/accuracy:** this PR's title, description, and test-plan describe `BinaryClassificationEvaluator.json`, `MulticlassClassificationEvaluator.json`, sample projects under `packages/uipath/samples/` (`binary_classification_agent/`, `multiclass_classification_simple/`), and an e2e test `test_classification_samples_e2e.py`.

None of these exist in the current diff (`gh api repos/UiPath/uipath-python/pulls/1663/files` — 13 files, none matching). The actual diff embeds `classes` + `aggregators` (precision/recall/fscore/confusion_matrix) directly into the existing `ExactMatchEvaluatorConfig` — a different, later design, confirmed by the commit history on this branch (`refactor(eval): move aggregators to ExactMatch, revert on Binary/Mult…`, `refactor(eval)!: trim PR to the ExactMatch aggregator contract only`). Consistent with that pivot, several review threads on this PR (from chatgpt-codex-connector and Copilot) are now anchored on files that no longer exist here — `evaluators_types/BinaryClassificationEvaluator.json` and `evaluators/multiclass_classification_evaluator.py`.

The design pivot itself looks like a genuine improvement (one aggregator implementation instead of two near-duplicate evaluator types) — this isn't asking to revert it. But as written:

- The test-plan's `cat packages/uipath/src/uipath/eval/evaluators_types/BinaryClassificationEvaluator.json` step can't be run — that file isn't in this PR.
- Anyone scanning merged PRs later will build the wrong mental model of what shipped (no new evaluator type IDs, no samples, no e2e — it's a `classes`/`aggregators` extension to `ExactMatch`, covered by unit tests only).
- The linked companion work (#1664, `UiPath/cli#2128`) may reference evaluator type IDs (`uipath-binary-classification`, `uipath-multiclass-classification`) that this PR no longer introduces — worth confirming those aren't now dangling.

Recommend updating the title/description/test-plan to describe what's actually shipping before merge.

---
*Posted by /deep-review (two-lens architecture + runtime review, slop-meter + mandatory Codex recheck). Two related code findings (F1: `negated` + `aggregators` gap, F2: out-of-vocab-as-skip) are posted as inline review comments on this PR.*

File: packages/uipath/src/uipath/eval/evaluators/exact_match_evaluator.py (hunk 11 above)
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/evaluators/exact_match_evaluator.py:60]: Fixed in 5b95d051 — added exactly this guard to `_validate_aggregators` (rejects `negated` + aggregators, mirroring the `case_sensitive` / `line_by_line_evaluator` guards). Locked by `test_rejects_aggregators_with_negated`. Thanks for the catch.

File: packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py (hunk 12 above)
Comment (ajay-kesavan) [reply packages/uipath/src/uipath/eval/evaluators/classification_dataset_evaluators.py]: Fixed in 5b95d051 — landed against this code path (not deferred). `_build_confusion` now counts a true-label datapoint whose *predicted* label is OOV as a false negative for the true class — a per-class `oov_fn` counter folded into `fn` / `support` / recall and the micro totals — instead of dropping it to `n_skipped`. Your worked example (9× `support→support` + 1× `support→fish`) now reports `support` recall **0.9** / F1 0.947 (was 1.0). Expected-label-OOV still skips: there's no in-vocab true class to attribute the miss to. `test_out_of_vocab_labels_are_skipped` is replaced by `test_out_of_vocab_prediction_counts_as_recall_miss` + `test_out_of_vocab_expected_label_is_skipped`.

**Review by Chibionos — APPROVED** [review]
Approving the SDK side. Reviewed the dataset-level classification evaluators (Precision / Recall / F-score / Confusion Matrix) end to end:

- Math is clean and the single source of truth — one `ClassificationDatasetEvaluator` consumed by both `uipath eval` locally and the platform reduce worker. No reimplementation, no drift.
- Confusion-matrix orientation (`[predicted][expected]`) is documented and pinned by a dedicated test.
- Config validation is thorough (classes required with aggregators, case-insensitive uniqueness, rejects line-by-line / case-sensitive combos) and the F-beta bound prevents NaN-on-the-wire.
- Strong edge-case coverage: empty input, out-of-vocab labels, malformed details, micro vs macro, dedup.

One note: land + publish this to PyPI before Agents#5307 merges, since that PR git-pins to this branch.
